### PR TITLE
Add `observeReadyElements` method

### PR DESCRIPTION
**Pull request by Richienb**
~~Depends on https://github.com/sindresorhus/type-fest/pull/246~~

___

~~I was unable to use `any-observable` for these reasons:~~
- ~~If I import it at the top level, the user will be forced to install an observable library even if they don't need it.~~
- ~~If I import it within the function, I will have to use the `import` function which will force me to make the observable-returning function a observable-resolving-promise-returning function which would not be user-friendly and a bit annoying to implement.~~
- ~~If I try to use `import` at the top-level in a try/catch block, I will need to require Node.js 14.8. ESLint also has trouble parsing this and making it work is quite inconvenient. https://github.com/eslint/eslint/issues/13178~~
- ~~`any-observable` doesn't use ESM and the interop layer Node.js uses doesn't seem to allow `import 'any-observable/register/zen';` to actually register the observer. https://github.com/sindresorhus/any-observable/issues/27~~

___

~~When running the unit tests locally, a test for a different part of code that this PR does not change failed:~~

https://github.com/sindresorhus/element-ready/blob/32360a00549b890e090f1c4649f6f30a89337758/test.js#L212

~~Difference:~~

```diff
- 'fulfilled'
+ 'pending'
```

~~This error may or may not reappear during runs of the GitHub Action.~~

___

Fixes https://github.com/sindresorhus/element-ready/issues/22

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#22: Subscribe to new elements on the page](https://issuehunt.io/repos/62587127/issues/22)
---
</details>

### CODE REVIEW

**Comment by sindresorhus** [comment]
Wouldn't it be better to use something more platform native and general like [async iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of)?

**Comment by Richienb** [comment]
> Wouldn't it be better to use something more platform native and general like [async iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for-await...of)?

Yes it would - changed the code.

**Comment by Richienb** [comment]
// @sindresorhus

**Comment by sindresorhus** [comment]
Instead of using `.stop()`, couldn't you just use `break` to stop the for await loop?

File: index.js [hunk 1]
```diff
@@ -1,5 +1,62 @@
 import ManyKeysMap from 'many-keys-map';
 import pDefer from 'p-defer';
+import Queue from 'yocto-queue';
+
+function createAsyncIterator() {
+	const queue = new Queue();
+	let isComplete = false;
+	const onNextCallbacks = new Set();
```
Comment (sindresorhus) [inline index.js:8]: I think all this should be in the iterator so it can be used multiple times.

File: index.js [hunk 2]
```diff
@@ -1,5 +1,62 @@
 import ManyKeysMap from 'many-keys-map';
 import pDefer from 'p-defer';
+import Queue from 'yocto-queue';
+
+function createAsyncIterator() {
```
Comment (sindresorhus) [inline index.js:5]: I think this creates an iterable, not an iterator.

File: index.js [hunk 3]
```diff
@@ -59,3 +116,69 @@ export default function elementReady(selector, {
...
+		handleMutations(observer.takeRecords());
+		observer.disconnect();
+		complete();
+	};
+
+	if (stopOnDomReady) {
+		target.addEventListener('DOMContentLoaded', () => {
+			console.log('loaded');
+			stop();
+		});
```
Comment (sindresorhus) [inline index.js:173]:
```suggestion
		}, {once: true});
```

File: index.js [hunk 4]
```diff
@@ -59,3 +116,69 @@ export default function elementReady(selector, {
...
+
+	const stop = () => {
+		handleMutations(observer.takeRecords());
+		observer.disconnect();
+		complete();
+	};
+
+	if (stopOnDomReady) {
+		target.addEventListener('DOMContentLoaded', () => {
+			console.log('loaded');
```
Comment (sindresorhus) [inline index.js:171]: Leftover

File: index.js [hunk 5]
```diff
@@ -59,3 +116,69 @@ export default function elementReady(selector, {
...
+	const {next, complete, iterator} = createAsyncIterator();
+
+	const handleMutations = mutations => {
+		for (const {addedNodes} of mutations) {
+			for (const element of addedNodes) {
+				if (element.nodeType !== 1) {
+					continue;
+				}
+
+				if (element.matches(selector)) {
```
Comment (sindresorhus) [inline index.js:135]: This could also be negated and use `continue`.

File: package.json [hunk 6]
```diff
@@ -39,7 +39,8 @@
 	"dependencies": {
 		"many-keys-map": "^1.0.3",
 		"p-defer": "^4.0.0",
-		"typed-query-selector": "^2.4.1"
+		"typed-query-selector": "^2.4.1",
+		"yocto-queue": "^0.1.0"
```
Comment (sindresorhus) [inline package.json:43]:
```suggestion
		"yocto-queue": "^1.0.0"
```

File: readme.md [hunk 7]
```diff
@@ -25,6 +25,19 @@ console.log(element.id);
 
 Returns a promise for a matching element.
 
+### observeReadyElements(selector, options?)
+
+Returns an async iterator which yields with each new matching element. Useful to for example, url shorten `<a>` tags as they are added or to click buttons when they are clickable.
```
Comment (sindresorhus) [inline readme.md:30]:
```suggestion
Returns an async iterator which yields with each new matching element. Useful to, for example, shorten the URLs of `<a>` tags as they are added or to click buttons when they are clickable.
```

Comment (sindresorhus) [reply readme.md:30]: Neither of these example use-cases are very common. I would try to come up with a more common use-case.

File: index.js [hunk 8]
```diff
@@ -1,5 +1,62 @@
...
+							};
+						}
+
+						if (isComplete) {
+							return {
+								done: true
+							};
+						}
+
+						await onNext();
```
Comment (sindresorhus) [inline index.js:49]:
```suggestion
						await onNext();

```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: readme.md (hunk 7 above)
Comment (Richienb) [reply readme.md:30]: Note to self:

> Useful in userscripts to modify elements when they are added or to wait for a specific amount of elements to load in

**Comment by Richienb** [comment]
> Instead of using `.stop()`, couldn't you just use `break` to stop the for await loop?

Indeed. Looks like as well as a `next()` method, I can also specify a [`return()` method](https://stackoverflow.com/a/42380518/8384910).

**Comment by sindresorhus** [comment]
Would be nice to have a test for `break` usage, and also show usage of `break` the docs for how to stop the loop, since it might not be immediately obvious to the user.

File: index.js (hunk 1 above)
Comment (Richienb) [reply index.js:8]: If an iterator is created for a second time, should it replay the values that the first one gave?

Comment (sindresorhus) [reply index.js:8]: No

**Comment by Richienb** [comment]
> Would be nice to have a test for `break` usage

As far as I can tell, there is no way to test if `break;` actually propogates to the corresponding internal logic but I'll still add it anyways in case of an unhandled promise exception.

**Comment by sindresorhus** [comment]
> As far as I can tell, there is no way to test if break; actually propogates to the corresponding internal logic but I'll still add it anyways in case it does and throws an error.

A test would at least confirm that it correctly stops and resolves with the correct items.

**Comment by Richienb** [comment]
CI is failing because of https://github.com/sindresorhus/p-state/pull/6

File: readme.md [hunk 9]
```diff
@@ -25,6 +25,23 @@ console.log(element.id);
 
 Returns a promise for a matching element.
 
+### observeReadyElements(selector, options?)
+
+Returns an async iterator which yields with each new matching element. Useful for user-scripts that modify elements when they are added.
```
Comment (sindresorhus) [inline readme.md:30]:
```suggestion
Returns an async iterable which yields with each new matching element. Useful for user-scripts that modify elements when they are added.
```

File: index.d.ts [hunk 10]
```diff
@@ -68,3 +67,32 @@ export default function elementReady<ElementName extends Element = HTMLElement>(
...
+```
+*/
+export function observeReadyElements<Selector extends string, ElementName extends Element = ParseSelector<Selector, HTMLElement>>(
+	selector: Selector,
+	options?: Options
+): AsyncIterable<ElementName>;
+export function observeReadyElements<ElementName extends Element = HTMLElement>(
+	selector: string,
+	options?: Options
+): AsyncIterableIterator<ElementName>;
```
Comment (sindresorhus) [inline index.d.ts:98]:
```suggestion
): AsyncIterable<ElementName>;
```

File: index.d.ts [hunk 11]
```diff
@@ -68,3 +67,32 @@ export default function elementReady<ElementName extends Element = HTMLElement>(
 	selector: string,
 	options?: Options
 ): StoppablePromise<ElementName | undefined>;
+
+/**
+Detect when elements are ready in the DOM. Returns an async iterator which yields with each new matching element. Useful for user-scripts that modify elements when they are added.
+
+@param selector - [CSS selector.](https://developer.mozilla.org/en-US/docs/Web/Guide/CSS/Getting_Started/Selectors) Prefix the element type to get a better return type. For example, `button.my-btn` instead of `.my-btn`.
+@returns An async iterator which yields with each new matching element.
```
Comment (sindresorhus) [inline index.d.ts:75]:
```suggestion
@returns An async iterable which yields with each new matching element.
```

File: index.d.ts [hunk 12]
```diff
@@ -68,3 +67,32 @@ export default function elementReady<ElementName extends Element = HTMLElement>(
 	selector: string,
 	options?: Options
 ): StoppablePromise<ElementName | undefined>;
+
+/**
+Detect when elements are ready in the DOM. Returns an async iterator which yields with each new matching element. Useful for user-scripts that modify elements when they are added.
```
Comment (sindresorhus) [inline index.d.ts:72]:
```suggestion
Detect when elements are ready in the DOM.

Useful for user-scripts that modify elements when they are added.
```

The return sentence is already below.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: index.js [hunk 13]
```diff
@@ -59,3 +60,68 @@ export default function elementReady(selector, {
...
+	stopOnDomReady = true,
+	waitForChildren = true,
+	timeout = Number.POSITIVE_INFINITY,
+} = {}) {
+	return {
+		[Symbol.asyncIterator]() {
+			const {next, complete, onCleanup, iterator} = createDeferredAsyncIterator();
+
+			function handleMutations(mutations) {
+				for (const {addedNodes} of mutations) {
```
Comment (sindresorhus) [inline index.js:75]: I think we could reduce this to one loop by running `.flatMap` on this first.

Comment (Richienb) [reply index.js:75]: `.flatMap()` does not work on NodeLists unless we use the spread operator which would result in a performance drop. This is the cleanest way.

**Comment by sindresorhus** [comment]
Regarding https://github.com/sindresorhus/element-ready/pull/36#discussion_r705933769. I disagree with your reasoning. The performance difference in real-world scenarios is negligible. But I also don't feel strongly about it and this PR has had enough back-and-forth, so I'm going to leave it.

**Comment by sindresorhus** [comment]
Thanks for contributing this :)

**Pull request merged**